packages/data/addon/builder/requestV2.ts

```diff
@@ -0,0 +1,176 @@
+  /**
+   * Convenience for adding a metric column
+   * @param metricName Name of metric
+   * @param parameters Dictionary of parameters
+   */
+  metric(metricName: string, parameters: Parameters = {}): RequestV2Builder {
+    this.column('metric', metricName, parameters);
+    return this;
+  }
+
+  /**
+   * Convenience for adding a dimension column
+   * @param dimensionName Name of dimension
+   * @param parameters Dictionary of parameters
+   */
+  dimension(dimensionName: string, parameters: Parameters = {}): RequestV2Builder {
+    this.column('dimension', dimensionName, parameters);
+    return this;
+  }
+
+  /**
+   * Adds a filter to the request
+   * @param type metric, dimension, or timeDimension
+   * @param field field name
+   * @param parameters dictionary of parameters
+   * @param operator operator
+   * @param values array of values for the filter
+   */
+  filter(
+    type: ColumnType,
+    field: string,
+    parameters: Parameters,
+    operator: FilterOperator,
+    values: (string | number | boolean)[]
+  ): RequestV2Builder {
+    this.request.filters.push({
+      type,
+      field,
+      parameters,
+      operator,
+      values
+    });
+    return this;
+  }
+
+  /**
+   * Adds a row limit to request
+   * @param limit number of max rows
+   */
+  limit(limit: number): RequestV2Builder {
+    this.request.limit = limit;
+    return this;
+  }
+
+  /**
+   * Adds sort to the request
+   * @param type metric,dimension or timeDimension
+   * @param field field name to sort by
+   * @param parameters Dictionary of parameters
+   * @param direction asc/desc
+   */
+  sort(type: ColumnType, field: string, parameters: Parameters, direction: SortDirection): RequestV2Builder {
+    this.request.sorts.push({
+      type,
+      field,
+      parameters,
+      direction
+    });
+    return this;
+  }
+
+  /**
+   * Adds table name to request
+   * @param name name of the table
+   */
+  table(name: string): RequestV2Builder {
+    this.request.table = name;
+    return this;
+  }
+
+  /**
+   * Allows specifies the datasource
+   * @param dataSource name of the datasource
+   */
+  dataSource(dataSource: string): RequestV2Builder {
+    this.request.dataSource = dataSource;
+    return this;
+  }
+
+  /**
+   * Convenience method for adding an interval filter
+   * @param grain 'day','week','month' etc... whatever your datasource supports
+   * @param start Period, datestring or macro
+   * @param end Period, datestring or macro
+   */
+  interval(grain: string, start: string, end: string): RequestV2Builder {
+    const table = this.request.table;
+    assert('Set the table before you set the interval', table.length !== 0);
+
+    this.column('timeDimension', `${table}.dateTime`, { grain }).filter(
+      'timeDimension',
+      `${table}.dateTime`,
+      { grain },
+      'bet',
+      [start, end]
+    );
+
+    return this;
+  }
+
+  /**
+   * Returns the built request
+   */
+  build() {
+    return this.request;
```

Comment:
Are we planning to add significant logic in `build`? If not, can we make this an object that implements `RequestV2` with the convenient methods? This way we can pass the object without calling `build()`.

Reply:
So the model is actually a bit overkill for this. The model comes with a lot of weight and dependencies that make building awkward, when all we need is a pojo that has the right interface to throw at the adapter.

Reply:
Also it wouldn't follow the builder pattern that way. The idea is to encapsulate the object while it's in an intermediate state as well as separate out building state concerns...

Reply:
When you say `model` are you speaking to the ember data model/fragment? I am not suggesting that.

The `build()` method, as is, doesn't offer anything. It is not doing any validation, optimization, or heavy lifting to build the object (like Java's StringBuilder).

>The intent of the Builder design pattern is to separate the construction of a complex object from its representation. By doing so, the same construction process can create different representations.

I am just not sure we are really gaining anything from keeping the request as internal state and then requiring the user to call `build`. We can still have method-chaining without a `build` method. 

If you really wanted a terminal method, it would be cool it you could just call `fetch` (like in JOOQ). Although, that is a little more complicated.

Reply:
Build simply does what most builders do, and unwraps the built object to be used. Basically it pulls the state out of the encapsulation. In most languages there would be a `new` here but in prototype languages that's building a pojo that's not really necessary. especially in TypeScript where we can check pojos against an interface like what we're doing here.

The gain we get is that we get an immutable state of each builder instance which is important for the use case as each dashboard usually creates a base query and then forks it in different ways to create a myriad of different queries.  I'm afraid without build() we would break that much needed encapsulation and make errors more likely.

I mean it kind of comes down to abandoning the builder pattern and replacing it with mutable class that is meant to be yet another implementation to requestV2. And also complicating the deep copy we have to do in the constructor.